src/mlpack/methods/ann/layer/multihead_attention_impl.hpp

```diff
@@ -0,0 +1,288 @@
+      embedDim, embedDim, false, false);
+
+  queryBias = arma::mat(weights.memptr()
+      + 4 * embedDim * embedDim, embedDim, 1, false, false).zeros();
+  keyBias = arma::mat(weights.memptr()
+      + (4 * embedDim + 1) * embedDim, embedDim, 1, false, false).zeros();
+  valueBias = arma::mat(weights.memptr()
+      + (4 * embedDim + 2) * embedDim, embedDim, 1, false, false).zeros();
+  outBias = arma::mat(weights.memptr()
+      + (4 * embedDim + 3) * embedDim, embedDim, 1, false, false).zeros();
+  XavierInitialization().Initialize(queryWt, embedDim, embedDim);
+  XavierInitialization().Initialize(keyWt, embedDim, embedDim);
+  XavierInitialization().Initialize(valueWt, embedDim, embedDim);
+  XavierInitialization().Initialize(outWt, embedDim, embedDim);
+}
+
+template <typename InputDataType, typename OutputDataType,
+          typename RegularizerType>
+template <typename eT>
+void MultiheadAttention<InputDataType, OutputDataType, RegularizerType>::
+Forward(const arma::Mat<eT>& input, arma::Mat<eT>& output)
+{
+  Log::Assert(input.n_rows == embedDim * (tLen + 2 * sLen));
+  const size_t bsz = input.n_cols;
+  output.set_size(embedDim * tLen, bsz);
+
+  arma::cube query(const_cast<arma::Mat<eT>&>(input).memptr(),
+      embedDim, tLen, bsz, false, false);
+  arma::cube key(const_cast<arma::Mat<eT>&>(input).memptr()
+      + query.n_elem, embedDim, sLen, bsz, false, false);
+  arma::cube value(const_cast<arma::Mat<eT>&>(input).memptr()
+      + query.n_elem + key.n_elem, embedDim, sLen, bsz, false, false);
+
+  for (size_t i = 0; i < bsz; ++i)
+  {
+    query.slice(i) = queryWt * query.slice(i) + arma::repmat(queryBias,1, tLen);
+    key.slice(i) = keyWt * key.slice(i) + arma::repmat(keyBias, 1, sLen);
+    value.slice(i) = valueWt * key.slice(i) + arma::repmat(valueBias, 1, sLen);
+  }
+  query /= std::sqrt(headDim);
+
+  if (!attnMask2d.is_empty())
+  {
+    if (attnMask2d.n_rows != sLen || attnMask2d.n_cols != tLen)
+      Log::Fatal << "The size of the 2D `attn_mask` is not correct."
+                 << std::endl;
+  }
+  else if (!attnMask3d.is_empty())
+  {
+    if (attnMask3d.n_rows != sLen || attnMask3d.n_cols != tLen ||
+        attnMask3d.n_slices != bsz * numHeads)
+      Log::Fatal << "The size of the 3D `attn_mask` is not correct."
+                 << std::endl;
+  }
+
+  if (!keyPaddingMask.is_empty())
+  {
+    if (keyPaddingMask.n_rows != sLen || keyPaddingMask.n_cols != 1)
+        Log::Fatal << "The size of the `keyPaddingMask` is not correct."
+                   << std::endl;
+  }
+
+  qProj = arma::cube(query.memptr(), headDim, tLen, numHeads * bsz, 0, 0);
+  kProj = arma::cube(key.memptr(), headDim, sLen, numHeads * bsz, 0, 0);
+  vProj = arma::cube(value.memptr(), headDim, sLen, numHeads * bsz, 0, 0);
+  arma::cube Wt3d(sLen, tLen, numHeads * bsz);
+
+  for (size_t i = 0; i < numHeads * bsz; ++i)
+  {
+    Wt3d.slice(i) = kProj.slice(i).t() * qProj.slice(i);
```

Comment:
Shouldn't it be
```c++
    Wt3d.slice(i) = qProj.slice(i) * kProj.slice(i).t();
```
?

Reply:
It's because I assume each embedding is a column vector and then any bunch of words when represented as embedding,
each column contains embedding for each word. In other implementations, it is other way round (i.e. each embedding is a row vector). I had to do this because 1) mlpack is column-major, 2) while splitting into n heads the actual embedding gets distorted.

Here, `kProj.slice(i)` is of shape `(headDim, sLen)` and `qProj.slice(i)` is of shape `(headDim, tLen)`, so
`qProj.slice(i) * kProj.slice(i).t()` won't be a valid multiplication.